sprint-1/step-2

DmitryMaximus opened the pull request.

DmitryMaximus committed:
sprint-1/step-2

ArslanMustafin commented on src/App.tsx:
```diff
@@ -1,26 +1,113 @@
+        getProductData()
+    }, [])
+
+    const [burger, setBurger] = useState({
+        top: null,
+        ingredients: [],
+        bottom: null
+    })
```
**Можно лучше:**

Лучше перенести стейт в `BurgerConstructor`. А также вместо `top` и `bottom`, используйте 1 поле, `bun`. Т.к. у нас булки одинаковые.

ArslanMustafin commented on src/App.tsx:
```diff
@@ -1,26 +1,113 @@
+
+    useEffect(() => {
+        const getProductData = async () => {
+            await fetch('https://norma.nomoreparties.space/api/ingredients ')
```
**Надо исправить:**

Если вы используйте asyc await, то не стоит использовать then, выберите что-то одно.

**Надо исправить:**

вынесите базовый адрес сервера `https://norma.nomoreparties.space/api` сервера в константу в отдельный файл, а адрес эндпоинта можно оставить в самой функции запроса `${BURGER_API_URL}/ingredients`

**Надо исправить**
Добавьте проверку ответа сервера - res.ok:

Например:

```
.then(res => res.ok ? res.json() : res.json().then((err) => Promise.reject(err));
```

иначе сейчас, если сервер ответит не кодом 2xx, а какой либо ошибкой (4xx, 5xx) мы не попадем в catch т.к. в этом случае fetch не выбрасывает ошибку, мы попытаемся отрисовать данные, хотя они не пришли, в итоге на попытке отрисовки несуществующих данных упадет все приложение

[https://learn.javascript.ru/fetch](https://learn.javascript.ru/fetch)

[https://developer.mozilla.org/en-US/docs/Web/API/Response/ok](https://developer.mozilla.org/en-US/docs/Web/API/Response/ok)

Т.к. эта проверка нужна у всех запросов, лучше вынести её в отдельную функцию

```js
const checkReponse = (res) => {
  return res.ok ? res.json() : res.json().then((err) => Promise.reject(err));
};
```

`fetch(someUrl).then(checkReponse)`
**Можно лучше**

в дальнейшем добавятся ещё запросы к бэкенду и лучше вынести описание этих запросов в функции

Например:

```js
function getIngredients() {
   return fetch(`${NORMA_API}/ingredients`)
    .then(проверка ответа)
}

```

и разместить их в отдельном файле  `utils/burger-api.js`, а не описывать запросы в самом компоненте

ArslanMustafin commented on src/App.tsx:
```diff
@@ -1,26 +1,113 @@
+    })
+
+
+    const [info, setInfo] = useState(false)
```
**Можно лучше:**

Данная переменная избыточно, т.к у мы можем просто смотреть, есть ли ingredient в `ingredient` или нет.

ArslanMustafin commented on src/App.tsx:
```diff
@@ -1,26 +1,113 @@
+    const [orderInfo, setOrderInfo] = useState(false)
+
+    useEffect(() => {
+        function handleEscapeKey(event: KeyboardEvent) {
```
**Надо исправить**

закрытие попапа по ESC относится к функционалу модального окна и должно быть описано в компоненте Modal, вешайте обработчик при открытии (монтировании) и удаляйте закрытии (размонтировании) применив useEffect

ArslanMustafin commented on src/App.tsx:
```diff
@@ -1,26 +1,113 @@
+            }
+        }
+
+        document.addEventListener('keydown', handleEscapeKey)
```
**Надо исправить**
Обработчик закрытия попапа должен добавляться при открытии попапа и удаляться при закрытии. Сейчас обработчик не будет удаляться при закрытии попапа т.к. из хука useEffect нужно возвращать колбэк отписки. Должно быть примерно так:

```js
   useEffect(() => {
      const handleESCclose = () => {............} //обработчик события
      document.addEventListener("keydown", handleESCclose); //навешиваем обработчик
	  
	  //возвращаем колбэк отписки, который  будет вызван при размонтировании компонента
      return () => document.removeEventListener("keydown", handleESCclose)                                                                                               
   }, [onClose])
```

ArslanMustafin commented on src/App.tsx:
```diff
@@ -1,26 +1,113 @@
+                    <BurgerConstructor handleClick={handleOpenOrderInfo} burger={burger}/>
+                </section>
+            </main>
+            <div style={{overflow: 'hidden'}}>
```
**Надо исправить:**

Не используйте инлайн стили, без необходимости, лучше сделайте дополнительный класс.

ArslanMustafin commented on src/App.tsx:
```diff
@@ -1,26 +1,113 @@
+            <div style={{overflow: 'hidden'}}>
+                {
+                    info &&
+                    <Modal onClose={handleClose} header={'Детали ингредиента'} children={<IngredientDetails item={ingredient}/>}/>
```
**Надо исправить:**

Здесь и далее.

Лучше использовать children следующим образом.

```js
<Modal onClose={handleClose} header={'Детали ингредиента'} >
  <IngredientDetails item={ingredient}/>
</Modal>
```

ArslanMustafin commented on src/App.tsx:
```diff
@@ -1,26 +1,113 @@
+
+    useEffect(() => {
+        if (data != null) {
+            const bun = data.filter((x: any) => x.type === "bun")[Math.random()<0.5?0:1];
```
**Можно лучше:**

используйте просто метод `find`, по типу `bun`.

ArslanMustafin commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,48 @@
+import styles from "./burger-constructor.module.css"
+import {Button, ConstructorElement, CurrencyIcon, DragIcon} from "@ya.praktikum/react-developer-burger-ui-components";
+
+const BurgerConstructor = ({ burger, handleClick }) => {
```
**Надо исправить:**

Для этого компонента, также требуется описать propTypes.

ArslanMustafin commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,48 @@
+
+const BurgerConstructor = ({ burger, handleClick }) => {
+
+    if (burger == null || burger.top == null || burger.bottom == null) return null
```
**Можно лучше:**

не нужно сравнивать столько всего.

Лучше в компоненте `App`, сделать 1 проверку.

```js
{data.length > 0 && (
    <section>
                    <BurgerIngredients handleClick={handleClick} data={data}/>
                </section>
                <section>
                    <BurgerConstructor handleClick={handleOpenOrderInfo} burger={burger}/>
                </section>
)}
```

ArslanMustafin commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,48 @@
+
+    if (burger == null || burger.top == null || burger.bottom == null) return null
+
+    const getSum = () => {
+        return burger.top.price + burger.bottom.price + burger.ingredients.map(x => x.price).reduce((partialSum, a) => partialSum + a, 0)
+    }
```
**Можно лучше:** 

Тут лучше использовать простую переменную вместе с `useMemo`, а не функцию).

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,70 @@
+        <div className={`mr-5 pt-10 ${styles.ingredients}`}>
+            <p style={{display: "flex"}} className="text text_type_main-large">Соберите бургер</p>
+            {/*табы*/}
+            <div className={'mt-5'} style={{display: 'flex'}}>
```
**Надо исправить:**

Здесь и во всём приложении.

Лучше не использовать инлайн стили.

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,70 @@
+            <div className={`${styles.list} pl-4 pr-4`}>
+                <div className={`${styles.setListWrapper} pt-10`}>
+                    {
+                        [...new Set(data.map(x => x.type))].map(x =>
```
**Можно лучше**

Лучше в компоненте делать filter по массиву, а потом отображать полученный массив.

Например:

```js
const buns = useMemo(data.filter((item) => item.type === 'bun'), [data]);
const mains = useMemo(data.filter((item) => item.type === 'main'), [data]);
const sauces = useMemo(data.filter((item) => item.type === 'sauce'), [data]);

return (
   ....................
   {buns.map(....................)}
   ..................
)
```

А не располагать фильтрацию в самом jsx

DmitryMaximus replied:
Я хотел уйти от хардкодинга  bun, main, sauce - поэтому старался сделать максимально гибкий компонент, есть ли другие варианты реализации?

ArslanMustafin replied:
Есть, можно из типов которые вам нужны сделать массив (и использовать include), и за практически 1 итерацию через Reduce создать 1 обьект, где поля - это тип ингредиента, а значение - массив.

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,70 @@
+                return "Начинки"
+            case "sauce":
+                return "Соусы"
+        }
```
**Можно лучше:**

Чтобы убрать варнинги от eslint, воспользуйтесь примером из документации. https://eslint.org/docs/latest/rules/default-case

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,70 @@
+import ModalOverlay from "../modal/modal-overlay";
+
+
+const BurgerIngredients = ({ data, handleClick }) => {
```
**Надо исправить:**

Для этого компонента, также следует описать propTypes.

ArslanMustafin commented on src/components/modal/ingredient-details.jsx:
```diff
@@ -0,0 +1,47 @@
+import styles from './ingredient-details.module.css'
+
+const IngredientDetails = ({ item }) => {
```
**Надо исправить:**

Для этого компонента, также следует описать propTypes.

ArslanMustafin commented on src/components/modal/modal-overlay.jsx:
```diff
@@ -0,0 +1,12 @@
+import styles from './model-overlay.module.css'
+import {useEffect} from "react";
+
+const ModalOverlay = ({ onClose }) => {
```
**Надо исправить:**

Для этого компонента, также следует описать propTypes.

ArslanMustafin commented on src/components/modal/modal.jsx:
```diff
@@ -0,0 +1,23 @@
+import ModalOverlay from "./modal-overlay";
+import {CloseIcon} from "@ya.praktikum/react-developer-burger-ui-components";
+
+const Modal = ({ children, header, onClose }) => {
```
**Надо исправить:**

Для этого компонента, также следует описать propTypes.

ArslanMustafin reviewed (changes requested):
## Code Review

Здравствуйте!

Отличная работа.

### **Вы молодец!**

- Теперь ингредиенты приходят от сервера;

- В приложении есть модальные окна;

### **Но есть несколько замечаний, которые нужно исправить:**

- Исправить ошибку в консоли: https://yadi.sk/d/5KUYHSfn5-IdLg;

- Очистить проект от остаточных файлов create-react-app: `App.test.tsx`.

- По чек-листу у вас, компонент `App` должен находиться в папочке `components/App`.

- Нужно поправить семантику в хедере, передалть кнопки в ссылки (оберните иконку и текст в ссылку). Только не забудьте проверить, чтобы активна была только 1 ссылка.

- У булок в компоненте `BurgerConstructor` по макету, должна быть приписка `(верх)` и `(низ)`.

- Что-то у вас слишком большая иконка, https://yadi.sk/d/1sldEbHQM-CVSw.

- Нужно описать propTypes у тех компонентов, где есть пропсы.

  также не забудьте, что нужно указывать `.isRequired`, без этого пропсы не будут проверяться.

  И требуется полное описание ингредиента (использовать `propTypes.shape()` и не забывайте про пункт выше),

  И последнее, вот корректная запись пропсов, где есть ингредиенты.

  ```js
  // для массива
  data: propTypes.arrayOf(ingredientType.isRequired).isRequired;
  // для 1 элемента
  ingredient: ingredientType.isRequired;
  ```

  Лучше всего будет вынести `ingredientType` в утилиты, и переиспользовать.

### **Сделает код лучше:**

- при клике на таб скролить список ингредиентов к нужному разделу.

- Вынесите в отдельную папку компоненты: `ingredients-detail`, `order-details`, `modal-overlay`.

### **_Подробно эти и более мелкие замечания, отмечены в коде._**

DmitryMaximus committed:
sprint-1/step-2 fix error "child should have a key"

DmitryMaximus committed:
sprint-1/step-2 react-app files delete

DmitryMaximus committed:
sprint-1/step-2
fix: app-header clickable link

DmitryMaximus committed:
sprint-1/step-2
fix: burger-constructor.jsx top/bottom naming render

DmitryMaximus committed:
sprint-1/step-2
fix: order-details.jsx fix tick image sizing

DmitryMaximus committed:
sprint-1/step-2
fix: PropTypes check added, project structure

DmitryMaximus committed:
sprint-1/step-2
fix: dynamic tab and blocks generation, hooks & refs, added default value in get label type function

ArslanMustafin commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,60 @@
+
+BurgerConstructor.propTypes = {
+    burger: propTypes.shape({
+        top: propTypes.shape({...ingredientType }),
```
**Надо исправить:**

Здесь и в других похожих местах.

Почему бы сразу не написать: 

А еще пропишите каждому полю `.isRequired`.
```js
export const ingredientType = PropTypes.shape({
    _id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    proteins: PropTypes.number.isRequired,
    fat: PropTypes.number.isRequired,
    carbohydrates: PropTypes.number.isRequired,
    calories: PropTypes.number.isRequired,
    price: PropTypes.number.isRequired,
    image: PropTypes.string.isRequired,
    image_mobile: PropTypes.string.isRequired,
    image_large: PropTypes.string.isRequired,
    __v: PropTypes.number.isRequired,
  });
 
  // И просто использовать 

  top: ingredientType,
```

ArslanMustafin commented on src/components/burger-constructor/burger-constructor.jsx:
```diff
@@ -0,0 +1,60 @@
+BurgerConstructor.propTypes = {
+    burger: propTypes.shape({
+        top: propTypes.shape({...ingredientType }),
+        ingredients: propTypes.arrayOf(propTypes.shape({...ingredientType })),
```
**Надо исправить:**

Не забывайте указывать тут `isRequired`. 

 ```js 
  // для массива
  data: propTypes.arrayOf(ingredientType.isRequired).isRequired;
```

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,85 @@
+        name: x,
+        ref: createRef(),
+        values: data.filter(z => z.type === x)
+    }))
```
**Можно лучше:** 

Оберните в useMemo, чтобы не было перерасчёта после ререндеров.

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,85 @@
+                        parsedData.map((x, index) =>
+                            <div key={index} className={styles.setList}>
+                                <p ref={x.ref} className={"text text_type_main-medium"}
+                                   style={{display: "flex"}}>{getTypeLabel(x.name)} </p>
```
**Надо исправить:**

Старайтесь не использовать инлайн стили, лучше создайте дополнительный класс.

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,85 @@
+                                            <p className={`${styles.name} text text_type_main-small mb-10`}>
+                                                {i.name}
+                                            </p>
+                                        </div>
```
**Можно лучше:**

Лучше вынести в отдельный компонент.

ArslanMustafin commented on src/components/burger-ingredients/burger-ingredients.jsx:
```diff
@@ -0,0 +1,85 @@
+
+BurgerIngredients.propTypes = {
+    data: PropTypes.arrayOf(PropTypes.shape({...ingredientType
+    })).isRequired,
```
**Надо исправить:**

Аналогичное замечание что и в `burger-constructor`.

ArslanMustafin commented on src/components/modal/ingredient-details/ingredient-details.jsx:
```diff
@@ -0,0 +1,47 @@
+import styles from './ingredient-details.module.css'
+
+const IngredientDetails = ({ item }) => {
```
**Надо исправить:**

Для этого компонента, нужно описать propTypes.

ArslanMustafin commented on src/components/modal/modal-overlay/modal-overlay.jsx:
```diff
@@ -0,0 +1,11 @@
+import styles from './model-overlay.module.css'
+
+const ModalOverlay = ({ onClose }) => {
```
**Надо исправить:**

Для этого компонента, нужно описать propTypes.

ArslanMustafin commented on src/components/App.tsx:
```diff
@@ -0,0 +1,113 @@
+import OrderDetails from "./modal/order-details/order-details";
+
+
+function App() {
```
**Надо исправить:**

Смените тип файла на `jsx`, в будущем вы будете использовать ts, но это ещё не скоро. А пока чтобы не мучаться с типами, лучше просто сменить расширении файлов, и убрать типы.

ArslanMustafin commented on src/components/App.tsx:
```diff
@@ -0,0 +1,113 @@
+    useEffect(() => {
+        function handleEscapeKey(event: KeyboardEvent) {
+            if (event.code === 'Escape') {
+                handleClose();
+                handleCloseOrderInfo();
```
**Надо исправить:**

Не стоит закрывать сразу все модальные окна таким образом, у вас есть 1 пропс `onClose`, вот в него и передавайте 1 функцию, которая будет закрывать только 1 модальное окно, а не все.

ArslanMustafin reviewed (changes requested):
## Code Review

Здравствуйте!

Отличная работа.

### **Вы молодец!**

- Практически все замечания с прошлого ревью были исправлены.

### **Но есть ещё несколько замечаний, которые нужно исправить:**

- У булок в компоненте `BurgerConstructor` по макету, должна быть приписка `(верх)` и `(низ)`. А у вас только `(низ)`. https://yadi.sk/d/1dGUI8T6zeTwpA;

- В папке `utils`, лежат только функции, поэтому там нужно поставить расширение `js`. Расширение `jsx`, мы используем там, где действительно есть `jsx` элементы.

- Нужно описать propTypes у тех компонентов, где есть пропсы.

- По чек-листу у вас, компонент `App` должен находиться в папочке `components/App`.
  Т.е. Компонент App должен тоже быть в папке, как и другие компоненты.

- Не исправлены прошлые замечания отмеченные в коде:

  - https://github.com/DmitryMaximus/react-burger/pull/2#discussion_r1039121508

  - https://github.com/DmitryMaximus/react-burger/pull/2#discussion_r1039115079

  - https://github.com/DmitryMaximus/react-burger/pull/2#discussion_r1039115558

  - https://github.com/DmitryMaximus/react-burger/pull/2#discussion_r1039115688

  - https://github.com/DmitryMaximus/react-burger/pull/2#discussion_r1039115947

  - https://github.com/DmitryMaximus/react-burger/pull/2#discussion_r1039116440

### **_Подробно эти и более мелкие замечания, отмечены в коде._**

DmitryMaximus committed:
sprint-1/step-2
fix: app file structure, missing propTypes, fix .jsx/.tsx formats

DmitryMaximus committed:
sprint-1/step-2
fix: changed top/bottom to bun, depreciated info and setInfo, inline styles

DmitryMaximus committed:
sprint-1/step-2
fix: changed api interaction

DmitryMaximus committed:
sprint-1/step-2
fix: added useMemo, inline styles removed

DmitryMaximus committed:
sprint-1/step-2
fix: burger-ingredients.jsx refactor, IngredientItem added

DmitryMaximus committed:
sprint-1/step-2
fix: fix modal window separate closing

ArslanMustafin commented on src/utils/api-url.js:
```diff
@@ -0,0 +1,2 @@
+
+export const burgerApiUrl = "https://norma.nomoreparties.space/api"
```
**Надо исправить**

имена констант принято записывать в стиле UPPER_CASE

ArslanMustafin reviewed (approved):
## Code Review

Здравствуйте!

Отличная работа.

### **Вы молодец! Работа принята, удачи в дальнейшем обучении.**

- Все замечания с прошлого ревью были исправлены.

- Отметил пару небольших замечаний в коде, которые рекомендую исправить.

### **Сделает код лучше:**

- https://github.com/DmitryMaximus/react-burger/pull/2#discussion_r1040419605, Лучше сделать так, а не использовать деструктуризацию каждый раз, когда вам нужно использовать описание ингредиента.

- Исправьте ошибки в консоли: https://yadi.sk/i/0-0ce8jzB1EwTQ;

### **_Подробно эти и более мелкие замечания, отмечены в коде._**

DmitryMaximus committed:
sprint-1/step-2
fix: fix url naming, propsTypes improved
